ocaml/xapi/xapi_vdi.ml

```diff
@@ -191,252 +145,302 @@ let check_operation_error ~__context ?sr_records:_ ?(pbd_records = [])
+      | `resize
+      | `resize_online
+      | `snapshot
+      | `set_on_boot ->
+          false
+      | `blocked
+      | `data_destroy
+      | `destroy
+      | `list_changed_blocks
+      | `force_unlock
+      | `forget
+      | `generate_config
+      | `update ->
+          true
+    in
+    let* () =
+      if
+        (not allowed_for_cbt_metadata_vdi)
+        && record.Db_actions.vDI_type = `cbt_metadata
+      then
         Error
-          ( Api_errors.operation_not_allowed
-          , [
-              "VDI containing HA statefile or redo log cannot be copied (check \
-               the VDI's allowed operations)."
-            ]
+          ( Api_errors.vdi_incompatible_type
+          , [_ref; Record_util.vdi_type_to_string `cbt_metadata]
           )
       else
         Ok ()
-  | `enable_cbt | `disable_cbt ->
-      if record.Db_actions.vDI_is_a_snapshot then
-        Error (Api_errors.operation_not_allowed, ["VDI is a snapshot: " ^ _ref])
-      else if not (List.mem record.Db_actions.vDI_type [`user; `system]) then
-        Error
-          ( Api_errors.vdi_incompatible_type
-          , [_ref; Record_util.vdi_type_to_string record.Db_actions.vDI_type]
-          )
-      else if reset_on_boot then
-        Error (Api_errors.vdi_on_boot_mode_incompatible_with_operation, [])
+    in
+    let allowed_when_cbt_enabled =
+      match op with
+      | `mirror | `set_on_boot ->
+          false
+      | `blocked
+      | `clone
+      | `copy
+      | `data_destroy
+      | `destroy
+      | `disable_cbt
+      | `enable_cbt
+      | `list_changed_blocks
+      | `force_unlock
+      | `forget
+      | `generate_config
+      | `resize
+      | `resize_online
+      | `snapshot
+      | `update ->
+          true
+    in
+    let* () =
+      if (not allowed_when_cbt_enabled) && record.Db_actions.vDI_cbt_enabled
+      then
+        Error (Api_errors.vdi_cbt_enabled, [_ref])
       else
         Ok ()
-  | `mirror
-  | `clone
-  | `generate_config
-  | `force_unlock
-  | `set_on_boot
-  | `list_changed_blocks
-  | `blocked
-  | `update ->
-      Ok ()
+    in
+    let check_destroy () =
+      if sr_type = "udev" then
+        Error (Api_errors.vdi_is_a_physical_device, [_ref])
+      else if is_tools_sr then
+        Error (Api_errors.sr_operation_not_supported, [Ref.string_of sr])
+      else if List.mem record.Db_actions.vDI_type [`rrd] then
+        Error (Api_errors.vdi_has_rrds, [_ref])
+      else if ha_enabled && vdi_is_ha_state_or_redolog then
+        Error (Api_errors.ha_is_enabled, [])
+      else if
+        vdi_is_ha_state_or_redolog
+        && Xapi_pool_helpers.ha_enable_in_progress ~__context
+      then
+        Error (Api_errors.ha_enable_in_progress, [])
+      else if
+        vdi_is_ha_state_or_redolog
+        && Xapi_pool_helpers.ha_disable_in_progress ~__context
+      then
+        Error (Api_errors.ha_disable_in_progress, [])
+      else
+        Ok ()
+    in
+    match op with
+    | `forget ->
```

Comment:
This could be encoded using `when condition ->`; it's a matter or taste. I like `when` because it avoids nesting.